### DISCUSSION TITLE
Documentation thread

### DISCUSSION THREAD

**Discussion by CremaFR** [LedgerHQ/ledger-fresh-management#62] [Specifications]
Issue to track all kind of documentation

[Excalidraw](https://excalidraw.com/#room=23577c2a1cc7e32a574b,vKLRiwt_E53x12RAexo46A) to explain wallet and plugin flow

**Comment by yogh333** [comment]
System Architecture:
https://app.excalidraw.com/l/19CgeSQLzX3/7EdSufWbeYJ

**Comment by yogh333** [comment]
I think some sequence diagrams detailing the following workflows (Fresh life cycle) are needed:

1. First time connection and Wallet deployment
2. Ledger plugin deployment for an account (WebAuthN, Starknet signer)
3. Non Ledger plugin deployment for an account (?)
4. Plugin removal (Ledger vs non-Ledger if any difference)
5. Plugin upgrade (Ledger vs non-Ledger if any difference)
6. Account upgrade
8. Wallet recovery

The workflows shall include all the involved actors/components i.e:

- Fresh Front End
- Fresh back ends: Plugin store, Web3 check, ...
- Starknet node
- Starknet Account contract
- Straknet Plugin contract
- Starknet Sequencer/Verifier (if needed)
- Ethereum node/contract (if needed)
- Fresh devices: Ledger Nanos/Stax, Mobile, Desktop, etc...

I propose to work on them, and will begin with the first one.

**Reply by imqdee** [reply]
Hi @yogh333!
I have taken the liberty of converting each element of the sequence diagrams list to a dedicated issue. All the issues created are tracked in this [project](https://github.com/orgs/LedgerHQ/projects/35).
I don't have created a task for elements 3 and 5 as I am not sure they are still relevant.

Feel free to update anything related to those tasks and the project created. Also, I assigned you to the tasks as you propose to work on them, feel free to unassign yourself if needed.

**Comment by yogh333** [comment]
[Deploy Fresh Account](https://lucid.app/lucidchart/28c9816f-92a9-451f-a919-87c7db925932/edit?invitationId=inv_8d097081-4256-43d1-994b-810fbfa5aadb)

Regarding this sequence, I have already several questions:

1. What kind of WebAuthn authenticators do we plan to support?
2. I have made the assumption that [Parsing and Validating the Registration Data](https://webauthn.guide/#about-webauthn) will be performed by the Deployer backend. Is it correct? 
3. When the Deployer backend shall deploy an account contract, does it rely on [UDC](https://community.starknet.io/t/universal-deployer-contract-proposal/1864)?
4. How is the first plugin (WenAuthn) deployed? I have made the assumption that it is the Deployer but how the Deployer, who has its own account contract, can add a plugin to an other account? On the other side, when an account is first deplyed without any plugin (i.e without any public key), how can we interact with it ?
5. Account address shall be recoverable as when the end user will reconnect later using the same WebAuthn authenticator, an new account shall not be deployed. How do we plan to store the account address (deterministic address?) ?

**Comment by rdubois-crypto** [comment]
A document of overview of the previous elements:
https://docs.google.com/document/d/1U9syluJr8WRNrY0pXDyrpW_LlG52x5vn2n4lJa_y-qo/edit#heading=h.pjkmb0bdpv6b

**Comment by yogh333** [comment]
> [Deploy Fresh Account](https://lucid.app/lucidchart/28c9816f-92a9-451f-a919-87c7db925932/edit?invitationId=inv_8d097081-4256-43d1-994b-810fbfa5aadb)
> 
> Regarding this sequence, I have already several questions:
> 
> 1. What kind of WebAuthn authenticators do we plan to support?
> 2. I have made the assumption that [Parsing and Validating the Registration Data](https://webauthn.guide/#about-webauthn) will be performed by the Deployer backend. Is it correct?
> 3. When the Deployer backend shall deploy an account contract, does it rely on [UDC](https://community.starknet.io/t/universal-deployer-contract-proposal/1864)?
> 4. How is the first plugin (WebAuthn) deployed? I have made the assumption that it is the Deployer but how the Deployer, who has its own account contract, can add a plugin to an other account? On the other side, when an account is first deplyed without any plugin (i.e without any public key), how can we interact with it ?
> 5. Account address shall be recoverable as when the end user will reconnect later using the same WebAuthn authenticator, an new account shall not be deployed. How do we plan to store the account address (deterministic address?) ?

Related to questions 3 and 4, it is yet not clear for me how the initial deployment is performed onchain. Below is a seq diagram describing on-chain operation, but I am not sure I have properly understood (DECLARE vs INVOKE tx, Deployer account vs Fresh account):

![Account Deployment (on chain)](https://user-images.githubusercontent.com/12996226/208084166-57fb24ce-1399-4625-acbd-f1d57fa41da7.png)

Would be happy to get your feedback !

**Comment by yogh333** [comment]
I have updated the system architecture diagram:
![image](https://user-images.githubusercontent.com/12996226/208400641-4ed1796f-381e-47ba-8c2a-f14afd5869d8.png)

General remark: this makes Fresh relies strongly on Ledger backends, which looks like more Web2.0 than Web3.0. Maybe I have missed some parts. Is it okay for everyone?

**Comment by yogh333** [comment]
Update: 
![image](https://user-images.githubusercontent.com/12996226/208788076-5c961288-2ba9-4895-b2dc-a2711f1f0951.png)

**Comment by yogh333** [comment]
Update: added Starknet RPC Provider
![image](https://user-images.githubusercontent.com/12996226/208885300-c1bca49c-87be-4ed1-abd3-3bc224b8361a.png)

**Comment by rdubois-crypto** [comment]
Architecture diagram of subcomponent, commands and connectors of the appfresh subsystem:
https://excalidraw.com/#json=KUMOODZQo_tJDZdxb7BXF,7aljKCKUJVS9otqjeiC2ww

![image](https://user-images.githubusercontent.com/103030189/208935174-2a093fac-ffcc-4466-9a74-ac0e4fa4d136.png)

**Comment by rdubois-crypto** [comment]
[Ledger Fresh Overview (1).pdf](https://github.com/LedgerHQ/ledger-fresh-management/files/10280295/Ledger.Fresh.Overview.1.pdf)

Current Document System Overview State

**Comment by yogh333** [comment]
Updated system architecture:
![image](https://user-images.githubusercontent.com/12996226/209156115-7bac015d-624f-458e-a25c-44362508b185.png)

**Comment by imqdee** [comment]
Updated system architecture:
<img width="1507" alt="image" src="https://user-images.githubusercontent.com/33158502/209190035-640b2cd6-fff9-48b6-83b5-d7603c904658.png">

This new version add the USB security key signer and add a dependency arrow between the modules "Signer vs Wallet" and "Notification center"

**Comment by rdubois-crypto** [comment]
Side note, do not hesitate to use larger font, tis hard to read without zooming when included in a doc.

**Comment by rdubois-crypto** [comment]
La présentation de Dan:
https://docs.google.com/presentation/d/19SkyPype0UMNlCEfp6Uc4WMpclawhCVGY1i2wCtYSx4/edit#slide=id.p

**Comment by imqdee** [comment]
Updated the system architecture.

<img width="1714" alt="image" src="https://user-images.githubusercontent.com/33158502/209366960-69019a63-cebc-4384-8db6-4e2173993024.png">

This new version renames the off-chain deployer to "deployment service", and introduces the on-chain deployer (our own deployer or the UDC -- will see).

@rdubois-crypto sorry dude, I didn't take the time this time to enlarge the font and as I'm stretching the architecture the fonts will look even smaller 😅

**Comment by imqdee** [comment]
Documents from this thread have been snapshotted and committed to the repository (3b15d68dbf13f6165d415699c99dc839c7edb708 6aaafce86db188de196bd1065353c9dd1c26f78c e9696d457c3395950552ae0048aa2e393b7dfc47). 
Keep in mind that from now, the version in the repository must be the latest, don't hesitate to update the document as much as needed.
